language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/json/JsonLfReader.java

```diff
@@ -0,0 +1,443 @@
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  public FromJson<BigDecimal> decimal() {
+    return () -> {
+      expectIsAt(
+          "decimal",
+          JsonToken.VALUE_NUMBER_INT,
+          JsonToken.VALUE_NUMBER_FLOAT,
+          JsonToken.VALUE_STRING);
+      BigDecimal value = null;
+      try {
+        value = new BigDecimal(parser.getText());
+      } catch (NumberFormatException e) {
+        parseExpected("decimal");
+      } catch (IOException e) {
+        parseExpected("decimal");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  public FromJson<Instant> timestamp() {
+    return () -> {
+      expectIsAt("timestamp", JsonToken.VALUE_STRING);
+      Instant value = null;
+      try {
+        value = Instant.parse(parser.getText());
+      } catch (DateTimeParseException e) {
+        parseExpected("timestamp");
+      } catch (IOException e) {
+        parseExpected("timestamp");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  public FromJson<LocalDate> date() {
+    return () -> {
+      expectIsAt("date", JsonToken.VALUE_STRING);
+      LocalDate value = null;
+      try {
+        value = LocalDate.parse(parser.getText());
+      } catch (DateTimeParseException e) {
+        parseExpected("date");
+      } catch (IOException e) {
+        parseExpected("date");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  public FromJson<String> party() {
+    return text();
+  }
+
+  public <C extends ContractId<?>> FromJson<C> contractId(Function<String, C> constr) {
+    return () -> {
+      String id = text().read();
+      return constr.apply(id);
+    };
+  }
+
+  public FromJson<String> text() {
+    return () -> {
+      expectIsAt("text", JsonToken.VALUE_STRING);
+      String value = null;
+      try {
+        value = parser.getText();
+      } catch (IOException e) {
+        parseExpected("valid textual value");
+      }
+      moveNext();
+      return value;
+    };
+  }
+
+  // Read an list with an unknown number of items of the same type.
+  public <T> FromJson<List<T>> list(FromJson<T> readItem) {
+    return () -> {
+      List<T> list = new ArrayList<>();
+      readStartArray();
+      while (notEndArray()) {
+        T item = readItem.read();
+        list.add(item);
+      }
+      readEndArray();
+      return list;
+    };
+  }
+
+  // Read a map with textual keys, and unknown number of items of the same type.
+  public <V> FromJson<Map<String, V>> textMap(FromJson<V> readValue) {
+    return () -> {
+      Map<String, V> map = new TreeMap<>();
+      readStartObject();
+      while (notEndObject()) {
+        String key = readFieldName();
+        V val = readValue.read();
+        map.put(key, val);
+      }
+      ;
+      readEndObject();
+      return map;
+    };
+  }
+
+  // Read a map with unknown number of items of the same types.
+  public <K, V> FromJson<Map<K, V>> genMap(FromJson<K> readKey, FromJson<V> readValue) {
```

Comment:
How about including a convenience instance for `Set`? This is really unpleasant to deal with manually. i.e. I mean the horrible
```json
{
  "a" : {},
  "b" : {}
} === Set("a", "b")
```

Reply:
Once code gen exists, the idea is that decoding sets would reduce to `Set.fromJson(json)`, where `Set` is the code-gen version of the one in the daml stdlib.

If we wanted native support for `Set`, beyond the daml library implementation, I think we'd want to make it a concept in the json lf encoding. There's also always the option of convenience functions to convert from the code-gen Set and a `java.util.Set` and/or `scala.collection.immutable.Set`